pkg/util/archive/zip.go

```diff
@@ -0,0 +1,261 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package archive provides functions to archive and unarchive files.
+package archive
+
+import (
+	"archive/zip"
+	"compress/flate"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+// Zip is an adapted implementation of (*Zip).Archive from
+// https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L140
+// Copyright (c) 2016 Matthew Holt
+func Zip(sources []string, destination string) error {
+	if !strings.HasSuffix(destination, ".zip") {
+		return fmt.Errorf("%s must have a .zip extension", destination)
+	}
+	if fileExists(destination) {
+		return fmt.Errorf("file already exists: %s", destination)
+	}
+	dir := filepath.Dir(destination)
+	if !fileExists(dir) {
+		err := os.MkdirAll(dir, 0755)
+		if err != nil {
+			return fmt.Errorf("%w making directory: %s", err, dir)
+		}
+	}
+
+	out, err := os.Create(destination)
+	if err != nil {
+		return fmt.Errorf("%w creating %s", err, destination)
+	}
+	defer out.Close()
+
+	zipW := zip.NewWriter(out)
+	zipW.RegisterCompressor(zip.Deflate, func(out io.Writer) (io.WriteCloser, error) {
+		return flate.NewWriter(out, flate.DefaultCompression)
+	})
+	defer zipW.Close()
+
+	for _, source := range sources {
+		err := writeWalk(zipW, source, destination)
+		if err != nil {
+			return fmt.Errorf("%w walking %s", err, source)
+		}
+	}
+
+	return nil
+}
+
+// UnZip unpacks the .zip file at source to destination.
+func UnZip(source, destination string) error {
+	destinationDir := filepath.Dir(destination)
+
+	if !fileExists(destinationDir) {
+		err := os.MkdirAll(destinationDir, 0755)
+		if err != nil {
+			return fmt.Errorf("preparing destination: %v", err)
+		}
+	}
+
+	zipR, err := zip.OpenReader(source)
+	if err != nil {
+		return fmt.Errorf("opening source file: %v", err)
+	}
+	defer zipR.Close()
+
+	for _, f := range zipR.File {
+		err := extractAndWriteFile(f, destination)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extractAndWriteFile(f *zip.File, destination string) error {
+	rc, err := f.Open()
+	if err != nil {
+		return fmt.Errorf("failed to open file: %v", err)
+	}
+	defer rc.Close()
+
+	path := filepath.Join(destination, f.Name)
+
+	// Check for ZipSlip (Directory traversal)
+	if !strings.HasPrefix(path, filepath.Clean(destination)+string(os.PathSeparator)) {
+		return fmt.Errorf("illegal file path: %s", path)
+	}
```

Comment:
This should probably use [`https://pkg.go.dev/github.com/cyphar/filepath-securejoin#SecureJoin`](https://pkg.go.dev/github.com/cyphar/filepath-securejoin#SecureJoin). If the linked implementation is trivial, it might be fine to inline.

Reply:
While I do agree that we might benefit from using a thrid party library for this functionality. The existing code works as expected and we have a test that validate that it works as expected. I would like to keep the `zip.go` as lean as possible 😄 

Unless @sgnn7 you feel strongly that we should use that third party library, I believe we can ship this at it is for now

Reply:
I think we do need this - we don't know who (or how) `Unzip` will be used in the future and the number of ways to escape out of basic path logic is surprisingly high. FWIW, the [code for securejoin](https://github.com/cyphar/filepath-securejoin/blob/main/join.go) is pretty slim and could in theory be inlined.

Reply:
Done https://github.com/DataDog/datadog-agent/pull/25498/commits/14c6666f62e44c6473ce3dcc0242e0e1ed463530